libkbfs/folder_branch_ops.go

```diff
@@ -791,6 +792,236 @@ func (fbo *folderBranchOps) startMonitorChat(tlfName tlf.CanonicalName) {
+	if err != nil {
+		return err
+	}
+
+	chs := make(map[string]<-chan struct{}, len(syncConfig.Paths))
+	// Look up and solo-sync each lead-up component of the path.
+pathLoop:
+	for _, p := range syncConfig.Paths {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		fbo.log.CDebugf(ctx, "Partially-syncing %s", p)
+
+		parentPath, syncedElem := stdpath.Split(p)
+		parents := strings.Split(strings.TrimSuffix(parentPath, "/"), "/")
+		currNode := rootNode
+		for _, parent := range parents {
+			if len(parent) == 0 {
+				continue
+			}
+			// TODO: parallelize the parent fetches and lookups.
+			currNode, _, err = fbo.Lookup(ctx, currNode, parent)
+			switch errors.Cause(err).(type) {
+			case NoSuchNameError:
+				fbo.log.CDebugf(ctx, "Synced path %s doesn't exist yet", p)
+				continue pathLoop
+			case nil:
+			default:
+				return err
+			}
+
+			// Use `PrefetchTail` for directories, to make sure that
+			// any child blocks in the directory itself get
+			// prefetched.
+			_, err = fbo.syncOneNode(
+				ctx, currNode, latestMerged, BlockRequestPrefetchTailWithSync)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Kick off a full deep sync of `syncedElem`.
+		elemNode, _, err := fbo.Lookup(ctx, currNode, syncedElem)
+		switch errors.Cause(err).(type) {
+		case NoSuchNameError:
+			fbo.log.CDebugf(ctx, "Synced element %s doesn't exist yet", p)
+			continue pathLoop
+		case nil:
+		default:
+			return err
+		}
+
+		ptr, err := fbo.syncOneNode(
+			ctx, elemNode, latestMerged, BlockRequestWithDeepSync)
+		if err != nil {
+			return err
+		}
+		ch, err := fbo.config.BlockOps().Prefetcher().
+			WaitChannelForBlockPrefetch(ctx, ptr)
+		if err != nil {
+			return err
+		}
+		chs[p] = ch
+	}
+
+	for p, ch := range chs {
+		select {
+		case <-ch:
+			fbo.log.CDebugf(ctx, "Prefetch for %s complete", p)
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
+func (fbo *folderBranchOps) kickOffPartialSync(
+	ctx context.Context, lState *lockState,
+	syncConfig keybase1.FolderSyncConfig, rmd ImmutableRootMetadata) {
+	// Kick off a background partial sync.
+	partialSyncCtx := fbo.ctxWithFBOID(context.Background())
+	fbo.log.CDebugf(
+		ctx, "Partial sync with a new context: FBOID=%s",
+		partialSyncCtx.Value(CtxFBOIDKey))
+	partialSyncCtx, cancel := context.WithCancel(partialSyncCtx)
+	fbo.partialSyncs.Add(1)
+	go func() {
+		defer cancel()
+		defer fbo.partialSyncs.Done()
+		_ = fbo.doPartialSync(partialSyncCtx, syncConfig, rmd)
+	}()
+
+	// Cancel the partial sync if the latest merged revision is updated.
+	updatedCh := func() <-chan struct{} {
+		fbo.headLock.Lock(lState)
+		defer fbo.headLock.Unlock(lState)
+		if rmd.Revision() != fbo.latestMergedRevision {
+			fbo.log.CDebugf(
+				partialSyncCtx, "Latest merged changed is now %d, not %d; "+
+					"aborting partial sync", fbo.latestMergedRevision,
+				rmd.Revision())
+			return nil
+		}
+		return fbo.latestMergedUpdated
```

Comment:
I don't think it can be `nil` once there's an `rmd` that can be passed to this method.  It's always set by `setLatestMergedRevisionLocked()`, which is called on all the MD initialization paths.